File: wechat-mp/scripts/media/upload_media.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
微信公众号素材上传工具
上传图片到素材库，获取 media_id 用于文章封面
"""

import argparse
import json
import os
import requests


def load_config(config_path: str = None):
    """加载配置文件"""
    if config_path is None:
        config_path = os.environ.get('WECHAT_MP_CONFIG') or os.path.expanduser('~/.config/wechat-mp/config.json')

    if not os.path.exists(config_path):
        print(f"配置文件不存在: {config_path}")
        return None

    with open(config_path, 'r', encoding='utf-8') as f:
        return json.load(f)


def get_access_token(appid: str, secret: str) -> str:
    """获取 access_token"""
    url = "https://api.weixin.qq.com/cgi-bin/token"
    params = {
        "grant_type": "client_credential",
        "appid": appid,
        "secret": secret
    }

    response = requests.get(url, params=params, timeout=30)
    data = response.json()

    if "access_token" in data:
        return data["access_token"]
    else:
        print(f"获取 access_token 失败: {data.get('errmsg', 'Unknown error')}")
        return None


def upload_image(image_path: str, access_token: str, permanent: bool = False) -> dict:
    """
    上传图片到素材库

    Args:
        image_path: 图片路径
        access_token: Access token
        permanent: 是否上传为永久素材

    Returns:
        dict: {"media_id": "xxx", "url": "xxx"} 或 {"error": "xxx"}
    """
    if not os.path.exists(image_path):
        print(f"文件不存在: {image_path}")
        return {"error": "File not found"}

    file_size = os.path.getsize(image_path)
    if permanent and file_size > 2 * 1024 * 1024:
        print(f"文件过大: {file_size / 1024 / 1024:.2f}MB，最大 2MB")
        return {"error": "File too large"}

    if permanent:
        url = f"https://api.weixin.qq.com/cgi-bin/material/add_material?access_token={access_token}&type=image"
    else:
        url = f"https://api.weixin.qq.com/cgi-bin/media/upload?access_token={access_token}&type=image"

    with open(image_path, 'rb') as f:
        files = {'media': f}
        response = requests.post(url, files=files, timeout=60)
        data = response.json()

    if "media_id" in data:
        return {
            "media_id": data["media_id"],
            "url": data.get("url", "")
        }
    else:
        print(f"上传失败: {data.get('errmsg', 'Unknown error')}")
        return {"error": data.get("errmsg", "Unknown error")}


def main():
    parser = argparse.ArgumentParser(description='微信公众号素材上传工具')
    parser.add_argument('--image', required=True, help='图片文件路径')
    parser.add_argument('--permanent', action='store_true',
                        help='上传为永久素材（封面必须用）')
    parser.add_argument('--config', help='配置文件路径')

    args = parser.parse_args()

    # 加载配置
    config = load_config(args.config)
    if not config:
        return

    wechat_config = config.get('wechat', {})
    appid = wechat_config.get('app_id')
    secret = wechat_config.get('app_secret')

    if not appid or not secret:
        print("配置文件中缺少 app_id 或 app_secret")
        return

    # 获取 access_token
    print("正在获取 access_token...")
    access_token = get_access_token(appid, secret)
    if not access_token:
        return
    print("access_token 获取成功")

    # 上传素材
    material_type = "永久" if args.permanent else "临时"
    print(f"正在上传 {material_type}素材: {args.image}")

    result = upload_image(args.image, access_token, args.permanent)

    if "media_id" in result:
        print(f"\n上传成功!")
        print(f"media_id: {result['media_id']}")
        if result.get("url"):
            print(f"url: {result['url']}")
    else:
        print(f"\n上传失败: {result.get('error')}")


if __name__ == '__main__':
    main()

File: wechat-mp/scripts/media/upload_media.mjs
#!/usr/bin/env node
// 微信公众号素材上传工具：上传图片到素材库，获取 media_id 用于文章封面。
// 零第三方依赖，使用 Node 18+ 内置 fetch / FormData / Blob。

import { readFileSync, existsSync, statSync } from 'node:fs';
import { basename } from 'node:path';
import { homedir } from 'node:os';

const TOKEN_URL = 'https://api.weixin.qq.com/cgi-bin/token';

function parseArgs(argv) {
  // 默认上传为永久素材（封面用途）；显式 --temp 才走临时素材（仅测试，3 天过期）。
  // --permanent 仍兼容（等价默认行为），无需再手动加。
  const args = { permanent: true };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === '--image') args.image = argv[++i];
    else if (a === '--config') args.config = argv[++i];
    else if (a === '--temp') args.permanent = false;
    else if (a === '--permanent') args.permanent = true;
  }
  return args;
}

function loadConfig(configPath) {
  const path = configPath
    || process.env.WECHAT_MP_CONFIG
    || `${homedir()}/.config/wechat-mp/config.json`;
  if (!existsSync(path)) {
    console.log(`配置文件不存在: ${path}`);
    return null;
  }
  return JSON.parse(readFileSync(path, 'utf-8'));
}

async function getAccessToken(appid, secret) {
  const url = `${TOKEN_URL}?grant_type=client_credential&appid=${encodeURIComponent(appid)}&secret=${encodeURIComponent(secret)}`;
  const data = await (await fetch(url, { signal: AbortSignal.timeout(30000) })).json();
  if (data.access_token) return data.access_token;
  console.log(`获取 access_token 失败: ${data.errmsg || 'Unknown error'}`);
  return null;
}

async function uploadImage(imagePath, accessToken, permanent) {
  if (!existsSync(imagePath)) {
    console.log(`文件不存在: ${imagePath}`);
    return { error: 'File not found' };
  }
  const size = statSync(imagePath).size;
  if (permanent && size > 2 * 1024 * 1024) {
    console.log(`文件过大: ${(size / 1024 / 1024).toFixed(2)}MB，最大 2MB`);
    return { error: 'File too large' };
  }
  const url = permanent
    ? `https://api.weixin.qq.com/cgi-bin/material/add_material?access_token=${accessToken}&type=image`
    : `https://api.weixin.qq.com/cgi-bin/media/upload?access_token=${accessToken}&type=image`;

  const form = new FormData();
  form.append('media', new Blob([readFileSync(imagePath)]), basename(imagePath));
  const data = await (await fetch(url, { method: 'POST', body: form, signal: AbortSignal.timeout(60000) })).json();

  if (data.media_id) return { media_id: data.media_id, url: data.url || '' };
  console.log(`上传失败: ${data.errmsg || 'Unknown error'}`);
  return { error: data.errmsg || 'Unknown error' };
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  if (!args.image) { console.error('缺少必填参数 --image'); process.exit(1); }

  const config = loadConfig(args.config);
  if (!config) return;

  const { app_id: appid, app_secret: secret } = config.wechat || {};
  if (!appid || !secret) { console.log('配置文件中缺少 app_id 或 app_secret'); return; }

  console.log('正在获取 access_token...');
  const accessToken = await getAccessToken(appid, secret);
  if (!accessToken) return;
  console.log('access_token 获取成功');

  console.log(`正在上传 ${args.permanent ? '永久' : '临时'}素材: ${args.image}`);
  const result = await uploadImage(args.image, accessToken, args.permanent);

  if (result.media_id) {
    console.log('\n上传成功!');
    console.log(`media_id: ${result.media_id}`);
    if (result.url) console.log(`url: ${result.url}`);
  } else {
    console.log(`\n上传失败: ${result.error}`);
  }
}

main();
